### PR TITLE
Validating hashes of uploaded file

### LINKED ISSUE #54: Checking checksums and sizes of uploaded files
`nginx-storage` should compute checksums of uploaded files while they're being uploaded. When the upload finishes, it should compare the checksum and file size with the ones stated in the token. If they don't match, reject the upload and return an error.

- `nginx-storage` doesn't have the means to decode golem messages. To work around it we'll have gatekeeper read the values from the token and return them in HTTP headers on the response (https://github.com/golemfactory/concent/issues/77). `nginx-storage` should read those values and use them in the comparison.
    - The headers are callled `Concent-File-Checksum` and `Concent-File-Size`.
    - The server should not read cheksum or file size from any headers on the request sent by the client. Only the ones from gatekeeper's response should be taken into account.
- The file should be moved to its final location only if both the checksum and file size match.
- If either the checksum or file size do not match, the file should be immediately deleted. The client should get a HTTP 400 error with a JSON message explaining why the file was rejected.
- No matter whether the operation succeeds or not, there should not be any leftover files after the operation (like for example temporary checksum files).
- Checksum must be computed incrementally, during upload, not in one operation afterwards. Otherwise we would block processing of other requests by the server.
- Checksum must begin with the name of the algorithm followed by a colon. e.g. `sha1:2bebc22296f03225617704d29d277c9e96fafcc2`. Checksums without algorithm name should not be accepted. The only algorithm supported right now is `sha1`. We'll add others later if they're needed.

### PR DESCRIPTION
**Pull request by bartoszbetka**
Resolves #54

### CODE REVIEW

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 1]
```diff
@@ -42,6 +48,22 @@ server {
                 return ngx.exit(400)
             end
 
+            -- Make sure that Concent-Checksum header was included with the request.
+            if ngx.req.get_headers()["Concent-Checksum"] == nil then
+                ngx.status = 400
+                ngx.header.content_type = "application/json; charset=utf-8"
+                ngx.say('{"error_message": "Concent-Checksum header is missing.", "error_code": "missing-header"}')
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: This should be HTTP 502. The error is not on client's side but on ours.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 2]
```diff
@@ -42,6 +48,22 @@ server {
...
+                return ngx.exit(400)
+            end
+
+            -- Make sure that Concent-File-Size header was included with the request.
+            if ngx.req.get_headers()["Concent-File-Size"] == nil then
+                ngx.status = 400
+                ngx.header.content_type = "application/json; charset=utf-8"
+                ngx.say('{"error_message": "Concent-File-Size header is missing.", "error_code": "missing-header"}')
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Same here. HTTP 502.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 3]
```diff
@@ -95,8 +122,42 @@ server {
             local utils = require('pl.utils')
             local table_utils   = require('table_utils')
 
-            data          = ngx.req.get_body_data()
-            tmp_file_name = string.match(data, 'id=(.*)&')
+            body_data                       = ngx.req.get_body_data()
+            tmp_file_name                   = string.match(body_data,'id=(.*)&')
+            sha1_uploaded_file              = string.match(body_data,'sha1=(.-)&')
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Fix your style here.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 4]
```diff
@@ -95,8 +122,42 @@ server {
             local utils = require('pl.utils')
             local table_utils   = require('table_utils')
 
-            data          = ngx.req.get_body_data()
-            tmp_file_name = string.match(data, 'id=(.*)&')
+            body_data                       = ngx.req.get_body_data()
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Call this `nginx_big_upload_request_body`.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 (hunk 3 above)
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: This variable is not really needed.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 5]
```diff
@@ -95,8 +122,42 @@ server {
...
+            tmp_file_name                   = string.match(body_data,'id=(.*)&')
+            sha1_uploaded_file              = string.match(body_data,'sha1=(.-)&')
+            sha1_uploaded_file_with_prefix  = "sha1:" .. sha1_uploaded_file
+            size_uploaded_file              = string.match(body_data,'size=(.-)&')
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Call this `file_size_from_nginx_big_upload`

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 6]
```diff
@@ -95,8 +122,42 @@ server {
...
+
+            -- Now that the file has successfully finished uploading, we can check if the hashes of file are the same.
+            -- Accept the name from Concent-Upload-Path header. We assume that this header was validated by
+            -- the gatekeeper app and is safe to use.
+            client_sha1 = ngx.req.get_headers()["Concent-Checksum"]
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Call this `checksum_from_gatekeeper`.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 7]
```diff
@@ -95,8 +122,42 @@ server {
...
+                os.remove("/srv/storage/" .. tmp_file_name)
+                os.remove("/srv/storage/" .. tmp_file_name .. ".shactx")
+                return ngx.exit(400)
+           end
+
+
+           -- Now that the file has successfully finished uploading, we can check if the size of file are the same.
+           -- Accept the name from Concent-File-Size header. We assume that this header was validated by
+           -- the gatekeeper app and is safe to use.
+           client_size= ngx.req.get_headers()["Concent-File-Size"]
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Call this `file_size_from_gatekeeper`.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 8]
```diff
@@ -95,8 +122,42 @@ server {
...
+           if  size_uploaded_file ~= client_size then
+                ngx.status = 400
+                ngx.header.content_type    = "application/json; charset=utf-8"
+                ngx.header["X-SHA1"]       = " "
+                ngx.header["X-Session-Id"] = " "
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Why a single space? Can't it be set to an empty string?

File: kubernetes/config-maps/nginx-storage/default.conf.j2 [hunk 9]
```diff
@@ -95,8 +122,42 @@ server {
             local utils = require('pl.utils')
             local table_utils   = require('table_utils')
 
-            data          = ngx.req.get_body_data()
-            tmp_file_name = string.match(data, 'id=(.*)&')
+            body_data                       = ngx.req.get_body_data()
+            tmp_file_name                   = string.match(body_data,'id=(.*)&')
+            sha1_uploaded_file              = string.match(body_data,'sha1=(.-)&')
+            sha1_uploaded_file_with_prefix  = "sha1:" .. sha1_uploaded_file
```
Comment (cameel) [inline kubernetes/config-maps/nginx-storage/default.conf.j2]: Call this `checksum_from_nginx_big_upload`.

File: kubernetes/config-maps/nginx-storage/default.conf.j2 (hunk 8 above)
Comment (bartoszbetka) [reply kubernetes/config-maps/nginx-storage/default.conf.j2]: With empty string not working, default header is return in response.

Comment (cameel) [reply kubernetes/config-maps/nginx-storage/default.conf.j2]: Then please add a comment explaining it. It's not obvious.

**Comment by cameel** [comment]
Please rebase the branch on master and squash the fixups.

**Pull request merged**